sdks/java/extensions/arrow/src/main/java/org/apache/beam/sdk/extensions/arrow/ArrowConversion.java

```diff
@@ -0,0 +1,448 @@
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Map type) {
+                    checkArgument(
+                        childrenFields.size() == 2,
+                        "Encountered "
+                            + childrenFields.size()
+                            + " child fields for map type, expected 2");
+                    return FieldType.map(
+                        toBeamField(childrenFields.get(0)).getType(),
+                        toBeamField(childrenFields.get(1)).getType());
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Int type) {
+                    if (!type.getIsSigned()) {
+                      throw new IllegalArgumentException("Unsigned integers are not supported.");
+                    }
+                    switch (type.getBitWidth()) {
+                      case 8:
+                        return FieldType.BYTE;
+                      case 16:
+                        return FieldType.INT16;
+                      case 32:
+                        return FieldType.INT32;
+                      case 64:
+                        return FieldType.INT64;
+                      default:
+                        throw new IllegalArgumentException(
+                            "Unsupported integer bit width: " + type.getBitWidth());
+                    }
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.FloatingPoint type) {
+                    switch (type.getPrecision()) {
+                      case SINGLE:
+                        return FieldType.FLOAT;
+                      case DOUBLE:
+                        return FieldType.DOUBLE;
+                      default:
+                        throw new IllegalArgumentException(
+                            "Unsupported floating-point precision: " + type.getPrecision().name());
+                    }
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Utf8 type) {
+                    return FieldType.STRING;
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Binary type) {
+                    return FieldType.BYTES;
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.FixedSizeBinary type) {
+                    return FieldType.logicalType(FixedBytes.of(type.getByteWidth()));
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Bool type) {
+                    return FieldType.BOOLEAN;
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Decimal type) {
+                    // FieldType.DECIMAL isn't perfect here since arrow decimal has a
```

Comment:
we the conversion is to Beam though is this really a problem?  or is the goal to only provide perfect round tripping?

Reply:
We might want to make DECIMAL a logical type at some point, parameterized by the precision.

Reply:
Yeah I actually lumped that into a jira for making DATETIME logical, [BEAM-7554](https://issues.apache.org/jira/browse/BEAM-7554). I'll add a comment referencing that here

Reply:
This is still a thorny issue, see this recent [dev@ thread](https://lists.apache.org/thread.html/r9cab743a2ec4278cc6195b624dc920fc42605b0a8966f1dc8058dcd2%40%3Cdev.beam.apache.org%3E). I'm inclined to save this for later as well.